### PR TITLE
plan: fix predicate push down for UnionScan

### PR DESCRIPTION
**Pull request by coocood**
### What problem does this PR solve? 

```
mysql> create table ta (a varchar(20));
Query OK, 0 rows affected (0.08 sec)

mysql> insert ta values ('1'), ('2');
Query OK, 2 rows affected (0.11 sec)

mysql> create table tb (a int);
Query OK, 0 rows affected (0.09 sec)

mysql> begin;
Query OK, 0 rows affected (0.00 sec)

mysql> insert tb values (0);
Query OK, 1 row affected (0.00 sec)

mysql> select * from ta where a = 1;
2 rows in set (0.00 sec)
```

Expected result is
```
1
```

Got
```
1
2
```

### What is changed and how it works?

The predicate push down for UnionScan plan returns retained predicates instead of nil.

If we return nil predicates, the parent Selection will be removed, but conditions in UnionScan is only used to evaluate dirty table added rows, the conditions that cannot be pushed down to coprocessor will be lost.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by coocood** [comment]
@jackysp @zz-jason @shenli PTAL

**Comment by zz-jason** [comment]
@winoros @eurekaka @XuHuaiyu PTAL

File: executor/union_scan_test.go [hunk 1]
```diff
@@ -78,3 +78,16 @@ func (s *testSuite) TestDirtyTransaction(c *C) {
 	tk.MustQuery("select * from t use index(idx) where c > 1 and d = 4").Check(testkit.Rows("1 2 3 4"))
 	tk.MustExec("commit")
 }
+
+func (s *testSuite) TestUnionScanWithCastCondition(c *C) {
```
Comment (zz-jason) [inline executor/union_scan_test.go]: It's better to add another explain test.

**Comment by winoros** [comment]
Eval it in `getSnapshotRow`?

**Comment by coocood** [comment]
@winoros 
I think do filter in Selection is better.

**Comment by coocood** [comment]
@zz-jason PTAL

File: cmd/explaintest/r/explain_easy.result [hunk 2]
```diff
@@ -352,3 +352,16 @@ create table t(a bigint primary key);
...
+drop table if exists ta, tb;
+create table ta (a varchar(20));
+create table tb (a varchar(20));
+begin;
+insert tb values ('1');
+explain select * from ta where a = 1;
+id	count	task	operator info
+Projection_5	8000.00	root	test.ta.a
+└─Selection_6	8000.00	root	eq(cast(test.ta.a), 1)
+  └─UnionScan_7	10000.00	root	eq(cast(test.ta.a), 1)
```
Comment (zz-jason) [inline cmd/explaintest/r/explain_easy.result]: why not `Eval it in getSnapshotRow`?

Comment (coocood) [reply cmd/explaintest/r/explain_easy.result]: It can be refactored later.

Comment (zz-jason) [reply cmd/explaintest/r/explain_easy.result]: The filter `a = 1`, which is `Selection_6`, should be added as a parent of `TableReader_9`.

Comment (coocood) [reply cmd/explaintest/r/explain_easy.result]: This is better, we can do it in the next PR.

**Pull request merged**